File: examples/stencil.jl

import cuNumeric

function initialize(N)
    println("Initializing stencil grid...")
    grid = cuNumeric.zeros(Float64, N + 2, N + 2)
    grid[:, 1] = -273.15
    grid[:, end] = -273.15
    grid[end, :] = -273.15
    grid[1, :] = 40.0
    return grid
end

function run_stencil(N, I, warmup)
    grid = initialize(N)

    println("Running Jacobi stencil...")

    center = grid[2:N+1, 2:N+1]
    north  = grid[1:N,   2:N+1]
    east   = grid[2:N+1, 3:N+2]
    west   = grid[2:N+1, 1:N]
    south  = grid[3:N+2, 2:N+1]

    for i in 1:(I + warmup)
        average = center .+ north .+ east .+ west .+ south
        work = 0.2 .* average
        center = work
    end
end


run_stencil(1000, 100, 5)

File: examples/stencil.py
#!/usr/bin/env python

# Copyright 2024 NVIDIA Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import cupynumeric as np


def initialize(N):
    print("Initializing stencil grid...")
    grid = np.zeros((N + 2, N + 2))
    grid[:, 0] = -273.15
    grid[:, -1] = -273.15
    grid[-1, :] = -273.15
    grid[0, :] = 40.0
    return grid


def run_stencil(N, I, warmup):  # noqa: E741
    grid = initialize(N)

    print("Running Jacobi stencil...")
    center = grid[1:-1, 1:-1]
    north = grid[0:-2, 1:-1]
    east = grid[1:-1, 2:]
    west = grid[1:-1, 0:-2]
    south = grid[2:, 1:-1]

    for i in range(I + warmup):
        average = center + north + east + west + south
        work = 0.2 * average
        center[:] = work


run_stencil(1000, 100, 5)
